packages/core/src/dialects/mssql/query-generator-internal.ts

```diff
@@ -45,4 +51,67 @@ export class MsSqlQueryGeneratorInternal<Dialect extends MssqlDialect = MssqlDia
 
     return fragment;
   }
+
+  getReturnFields(options: GetReturnFieldsOptions, modelDefinition?: ModelDefinition | null): string[] {
+    const returnFields: string[] = [];
+
+    if (Array.isArray(options.returning)) {
+      returnFields.push(...options.returning.map(field => {
+        if (typeof field === 'string') {
+          return this.queryGenerator.quoteIdentifier(field);
+        } else if (field instanceof Literal) {
+          // Due to how the mssql query is built, using a literal would never result in a properly formed query.
+          // It's better to warn early.
+          throw new Error(`literal() cannot be used in the "returning" option array in ${this.dialect.name}. Use col(), or a string instead.`);
+        } else if (field instanceof BaseSqlExpression) {
+          return this.queryGenerator.formatSqlExpression(field, options);
+        }
+
+        throw new Error(`Unsupported value in "returning" option: ${inspect(field)}. This option only accepts true, false, an array of strings, or a sql expression.`);
+      }));
+    } else if (modelDefinition) {
+      for (const [name, attribute] of modelDefinition.physicalAttributes) {
+        returnFields.push(this.queryGenerator.quoteIdentifier(attribute.columnName ?? name));
+      }
+    }
+
+    if (returnFields.length === 0) {
+      returnFields.push('*');
+    }
+
+    return returnFields;
+  }
+
+  generateOutputTableFragment(returnFields: string[], modelDefinition?: ModelDefinition | null): string {
+    if (!(modelDefinition instanceof ModelDefinition)) {
+      throw new Error('Using returning with triggers requires specifying a model or model definition.');
+    }
+
+    const columnsToExclude = new Set<string>();
+    const attributes = map(filter(returnFields, field => field !== '*'), field => {
+      const unquotedColumn = field.replace(/^\[/, '').replace(/\]$/, '');
```

Comment:
Quoting then unquoting is a bad idea and will break for identifiers that require escaping

The issue is that `getReturnFields` handles both formatting `returning`, & setting its default value.

You need to first use a function called `normalizeReturning` which:

- if `returning` is true: sets it to the list of model columns when a model is available, or `*` otherwise
- if `returning` is false: sets it to `EMPTY_ARRAY`
- if `returning` is an array: leaves it as-is

Then `getReturnFields` can become `formatReturnField`, which only handles serialization

Reply:
I agree and have updated the implementation now.